<a href="https://colab.research.google.com/github/Nasriiin/Nasriiin/blob/main/Nasrin_Soleimani_Python_Midterm_Test.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


![titanic.jpg](https://image.geo.de/31780898/t/gt/v3/w960/r1.5/-/titanic---pa--122855442.jpg)
# Midterm Test 

Hello!

You already learned a lot how to explore and understand data. Let's test this! If you hand in your solution we will check it so you get direct feedback and know where you stand. To get feedback save a copy and rename it, so it contains your name. Then save the copy to this 
[folder](https://drive.google.com/drive/folders/1JqOTlRt9VhK3NVHt0svBHLI8DnMdpp5Q?usp=sharing).

The deadline for handing in your solution is May 9.

You can answer those questions as you wish by just coding or inspecting the data through a plot. But please answer the question with a sentence or two where you explain what the result is telling you. 

If you have any questions, please ask us on slack. We are happy to help.

Have fun coding!

### Titanic Dataset

Today we will explore a new dataset, the infamous Titanic dataset. It contains informations on passengers of the Titanic. At kaggle many people already investigated this dataset. Check it out to get new ideas: -> https://www.kaggle.com/c/titanic/overview


The dataset contains 12 columns.

- survival: 	Survival 	0 = No, 1 = Yes
- pclass: 	Ticket class 	1 = 1st, 2 = 2nd, 3 = 3rd
- sex: 	Sex 	
- Age: 	Age in years 	
- sibsp: 	# of siblings / spouses aboard the Titanic 	
- parch: 	# of parents / children aboard the Titanic 	
- ticket: 	Ticket number 	
- fare: 	Passenger fare 	
- cabin: 	Cabin number 	
- embarked: 	Port of Embarkation 	C = Cherbourg, Q = Queenstown, S = Southampton

It can be downloaded from: https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv


### 1. Load Libraries and data

Load libraries you might need and load the data using pandas. Get a first overview:
- Is the data loaded correctly?
- How does it look like?
- How many rows, how many columns?
- Do we see any Null values? What could they mean?

In [ ]:
import pandas as pd
import seaborn as sns
import numpy as np

In [ ]:
df = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")


In [ ]:
df.head()



PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
df.shape
#the dataset has 891 rows and 12 columns.

(891, 12)

In [ ]:
df.isna().sum()
#there is null values in age columns,cabin, and embarked as well.

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

### 2. What is the average fare per class? On which range do the fares per 

1.   List item
2.   List item

class spread?

In [ ]:
df_class_2=df[df['Pclass']==2]
df_class_2.head(10)
print(df_class_2['Fare'].mean())
print(df_class_2['Fare'].min())
print(df_class_2['Fare'].max())

#The average class

20.66218315217391
0.0
73.5


The average far per class 1 is 20.662 and the range is between 0 and **73.5**

In [ ]:
df_class_3=df[df['Pclass']==3]
df_class_3.head(10)
print(df_class_3['Fare'].mean())
print(df_class_3['Fare'].min())
print(df_class_3['Fare'].max())

13.675550101832997
0.0
69.55


The average far per class 1 is 13.675 and the range is between 0 and 69.55

In [ ]:
df_class_1=df[df['Pclass']==1]
df_class_1.head(10)
print(df_class_1['Fare'].mean())
print(df_class_1['Fare'].min())
print(df_class_1['Fare'].max())

84.15468749999992
0.0
512.3292


The average far per class 1 is 84.154 and the range is between 0 and 512.3292

### 3. How many people had a cabin? 

In [ ]:
df['Cabin'].isna().sum()
#the number of null values is 687 so the we can extract the people that had a cabin by calculate the minus between the number of rows in dataset and 687

687

the number of null values is 687 so the we can extract the people that had a cabin by calculate the minus between the number of rows in dataset and 687 
so the number of people that had a cabin are 891 - 687 = 204

4. At which port lived the richest?




In [ ]:
df_class_1['Embarked'].value_counts()

S    127
C     85
Q      2
Name: Embarked, dtype: int64

so we can extract from the count  of values in column Embearked that the S value is the highest value the most of the richest people that choosed first calss were living in Southampton 

### 5. What is the most frequent last name?


In [ ]:
df["Name"].str.lower()
df["Name"].str.split(",")
df["Surname"] = df["Name"].str.split(",").str.get(0)
df["Surname"]
df.head()


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked    Surname  
0      0         A/5 21171   7.2500   NaN        S     Braund  
1      0          PC 17599  71.2833   C85        C    Cumings  
2      0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen  
3      0            113803  53.1000  C123        S   Futrelle  
4      0            373450   8.0500   NaN        S      Allen

In [ ]:

df['Surname'].mode()


0    Andersson
dtype: object

### 6. Not for every person the age is noted. Fill in the gaps with the average age per gender.

In [ ]:
#Finding the mean of the column having NaN
mean_value=df['Age'].mean()
print(mean_value)

29.69911764705882


In [ ]:
df['Age'].fillna(value=mean_value, inplace=True)
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
Surname          0
dtype: int64

### Bonus: Upload this file as .ipnyb - notebook in your personal git-hub account and share the address here.

In [ ]:
# address: